start_admin_sshd.sh

```diff
@@ -4,64 +4,122 @@
 set -e
 
-mkdir -p /home/ec2-user/.ssh/
-chmod 700 /home/ec2-user/.ssh/
-ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
-ssh_config_dir="/home/ec2-user/.ssh"
-
-# Populate authorized_keys with all the public keys found in instance meta-data
-# The URLs for keys include an index and the keypair name, e.g.
-# http://169.254.169.254/latest/meta-data/public-keys/0=mykeypair/openssh-key
-ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
-touch ${ssh_authorized_keys}
-chmod 600 ${ssh_authorized_keys}
-public_key_base_url="http://169.254.169.254/latest/meta-data/public-keys/"
-imds_session_token=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 60")
-imds_request_add_header="X-aws-ec2-metadata-token: ${imds_session_token}"
-public_key_indexes=($(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}" \
-    | cut -d= -f1 \
-    | xargs))
-
-for public_key_index in "${public_key_indexes[@]}"; do
-  public_key_data="$(curl -H "${imds_request_add_header}" -sf "${public_key_base_url}/${public_key_index}/openssh-key")"
-  if [[ ! "${public_key_data}" =~ ^"ssh" ]]; then
-    echo "Key ${public_key_data} with index ${public_key_index} looks invalid" >&2
-    continue
-  fi
-  echo "${public_key_data}" >> "${ssh_authorized_keys}"
-  if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
-    echo "Failed to write key with index ${public_key_index} to authorized_keys" >&2
-    continue
-  fi
-done
+log() {
+  echo "$*" >&2
+}
 
-# If we didn't write any keys at all, there's not much point in continuing
-if [ ! -s "${ssh_authorized_keys}" ]; then
-  echo "Failed to write any valid public keys to authorized_keys" >&2
-  exit 1
+declare -r ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
+declare -r user_data="/.bottlerocket/host-containers/admin/user-data"
+declare -r user_ssh_dir="/home/ec2-user/.ssh"
+available_auth_methods=0
+
+mkdir -p "${user_ssh_dir}"
+chmod 700 "${user_ssh_dir}"
+
+parse_authorized_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["authorized_keys"][]' "${user_data}" 2>/dev/null); then
+      log "Failed to parse authorized_keys from ${user_data}"
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        log "Failed to validate ${key}"
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate authorized_keys with all the authorized keys found in user-data
+if authorized_keys=$(parse_authorized_keys); then
+  ssh_authorized_keys="${user_ssh_dir}/authorized_keys"
+  install -D -m 0600 /dev/null "${ssh_authorized_keys}"
```

Comment:
This results in `~/.ssh` getting created with the default umask, but the recommended default is `0700`.